ssh/tailssh/tailssh.go

```diff
@@ -208,98 +240,63 @@ func (srv *server) handleAcceptedSSH(ctx context.Context, s ssh.Session, ci *ssh
 	logf := srv.logf
 	localUser := lu.Username
 
-	var err error
-	ptyReq, winCh, isPty := s.Pty()
-	logf("ssh: connection from %v %v to %v@ => %q. command = %q, env = %q", ci.srcIP, ci.uprof.LoginName, ci.sshUser, localUser, s.Command(), s.Environ())
-	var cmd *exec.Cmd
 	if euid := os.Geteuid(); euid != 0 {
 		if lu.Uid != fmt.Sprint(euid) {
 			logf("ssh: can't switch to user %q from process euid %v", localUser, euid)
 			fmt.Fprintf(s, "can't switch user\n")
 			s.Exit(1)
 			return
 		}
-		cmd = exec.Command(loginShell(lu.Uid))
-		if rawCmd := s.RawCommand(); rawCmd != "" {
-			cmd.Args = append(cmd.Args, "-c", rawCmd)
-		}
-	} else {
-		if rawCmd := s.RawCommand(); rawCmd != "" {
-			cmd = exec.Command("/usr/bin/env", "su", "-c", rawCmd, localUser)
-			// TODO: and Env for PATH, SSH_CONNECTION, SSH_CLIENT, XDG_SESSION_TYPE, XDG_*, etc
-		} else {
-			cmd = exec.Command("/usr/bin/env", "su", "-", localUser)
-		}
 	}
-	cmd.Dir = lu.HomeDir
-	cmd.Env = append(cmd.Env, s.Environ()...)
-	cmd.Env = append(cmd.Env, envForUser(lu)...)
-	if ptyReq.Term != "" {
-		cmd.Env = append(cmd.Env, fmt.Sprintf("TERM=%s", ptyReq.Term))
-	}
-	logf("Running: %q", cmd.Args)
-	var toCmd io.WriteCloser
-	var fromCmd io.ReadCloser
-	if isPty {
-		f, err := pty.StartWithSize(cmd, &pty.Winsize{
-			Rows: uint16(ptyReq.Window.Width),
-			Cols: uint16(ptyReq.Window.Height),
-		})
+
+	cmd, stdin, stdout, stderr, err := srv.launchProcess(ctx, s, ci, lu)
+	if err != nil {
+		logf("start failed: %v", err.Error())
+		s.Exit(1)
+		return
+	}
+	// We use this sync.Once to ensure that we only terminate the process once,
+	// either it exits itself or is terminated
+	var exitOnce *sync.Once
+	if ctx.Done() != nil {
+		exitOnce = &sync.Once{}
```

Comment:
I'd just make Line 260 a var and use &exitOnce in places, but sure, this works too.